core/src/main/java/tech/pegasys/web3signer/core/multikey/metadata/yubihsm/YubiHsmOpaqueDataProvider.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.web3signer.core.multikey.metadata.yubihsm;
+
+import tech.pegasys.signers.yubihsm.pkcs11.Pkcs11Module;
+import tech.pegasys.signers.yubihsm.pkcs11.Pkcs11Session;
+import tech.pegasys.signers.yubihsm.pkcs11.Pkcs11YubiHsm;
+import tech.pegasys.signers.yubihsm.pkcs11.Pkcs11YubiHsmPin;
+import tech.pegasys.web3signer.core.multikey.metadata.YubiHsmSigningMetadata;
+
+import java.nio.file.Path;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.apache.tuweni.bytes.Bytes;
+
+// Note: PKCS11 Sun provider has a limitation that one loaded module can only connect to one YubiHSM
+// device
+// In order to connect to multiple YubiHSM devices (in parallel), different module
+// installations/path must be used.
+public class YubiHsmOpaqueDataProvider implements AutoCloseable {
+  // maintains a cache of YubiHSM PKCS11 module and session to avoid module/session duplication
+  private final Map<String, Pkcs11Module> pkcs11ModuleMap = new ConcurrentHashMap<>();
+  private final Map<String, Pkcs11Session> pkcs11SessionMap = new ConcurrentHashMap<>();
```

Comment:
wondering about the string key-type - its appears arbitrary - is it worth having an object which represents the key - then rely on Java hashcode to ensure equality etc?

Reply:
maybe even just use the metadata object? Alternatively, have a "set" rather than a map?

Reply:
The session String 'identifier' is a combination of `modulePath + connectorUrl + AuthId`. We can create a separate value object to represent the same (in fact I had one earlier, then removed it as it was easily constructed using a method within this class). It has to be a map because we have to look up already created sessions.